File: hypersim-sdk/src/security/RateLimiter.ts
/**
 * Advanced rate limiting with DDoS protection
 */

import { EventEmitter } from 'events';
import { RateLimitConfig } from './types';

interface RateLimitEntry {
  count: number;
  window: number;
  queue: Array<{ request: any; resolve: Function; reject: Function; timestamp: number }>;
  suspiciousActivity: number;
  lastRequest: number;
}

interface DDoSMetrics {
  requestsPerSecond: number;
  uniqueIPs: Set<string>;
  suspiciousPatterns: number;
  blockedRequests: number;
}

export class RateLimiter extends EventEmitter {
  private readonly config: Required<RateLimitConfig>;
  private readonly limits: Map<string, RateLimitEntry> = new Map();
  private readonly ddosMetrics: DDoSMetrics;
  private cleanupInterval: NodeJS.Timeout;

  constructor(config: RateLimitConfig) {
    super();
    
    this.config = {
      requestsPerWindow: config.requestsPerWindow,
      windowMs: config.windowMs,
      maxQueueSize: config.maxQueueSize ?? 100,
      ddosProtection: config.ddosProtection ?? true,
      ddosThreshold: config.ddosThreshold ?? 1000
    };

    this.ddosMetrics = {
      requestsPerSecond: 0,
      uniqueIPs: new Set(),
      suspiciousPatterns: 0,
      blockedRequests: 0
    };

    // Cleanup old entries every minute
    this.cleanupInterval = setInterval(() => {
      this.cleanup();
    }, 60000);

    // DDoS monitoring every second
    if (this.config.ddosProtection) {
      setInterval(() => {
        this.monitorDDoS();
      }, 1000);
    }
  }

  /**
   * Check if request is within rate limit
   */
  async checkLimit(request: any): Promise<boolean> {
    const identifier = this.getIdentifier(request);
    const now = Date.now();
    
    // Check for DDoS patterns
    if (this.config.ddosProtection && this.isDDoSRequest(request, identifier)) {
      this.ddosMetrics.blockedRequests++;
      this.emit('ddosDetected', {
        identifier,
        timestamp: now,
        metrics: this.getDDoSMetrics()
      });
      return false;
    }

    let entry = this.limits.get(identifier);
    
    if (!entry) {
      entry = {
        count: 0,
        window: now,
        queue: [],
        suspiciousActivity: 0,
        lastRequest: 0
      };
      this.limits.set(identifier, entry);
    }

    // Reset window if expired
    if (now - entry.window >= this.config.windowMs) {
      entry.count = 0;
      entry.window = now;
      entry.suspiciousActivity = 0;
    }

    // Check for suspicious patterns
    const timeSinceLastRequest = now - entry.lastRequest;
    if (timeSinceLastRequest < 100) { // Less than 100ms between requests
      entry.suspiciousActivity++;
      if (entry.suspiciousActivity > 10) {
        this.emit('suspiciousActivity', {
          identifier,
          pattern: 'rapid_requests',
          count: entry.suspiciousActivity
        });
        return false;
      }
    }

    entry.lastRequest = now;

    // Check rate limit
    if (entry.count >= this.config.requestsPerWindow) {
      // Queue request if queueing is enabled
      if (this.config.maxQueueSize > 0 && entry.queue.length < this.config.maxQueueSize) {
        return new Promise((resolve, reject) => {
          entry!.queue.push({
            request,
            resolve,
            reject,
            timestamp: now
          });
          
          // Process queue when window resets
          setTimeout(() => {
            this.processQueue(identifier);
          }, this.config.windowMs - (now - entry!.window));
        });
      }
      
      this.emit('rateLimitExceeded', {
        identifier,
        count: entry.count,
        window: entry.window,
        timestamp: now
      });
      
      return false;
    }

    entry.count++;
    this.updateDDoSMetrics(request, identifier);
    return true;
  }

  /**
   * Get current rate limit status for identifier
   */
  getStatus(identifier: string): {
    count: number;
    limit: number;
    remaining: number;
    resetTime: number;
    queueLength: number;
  } {
    const entry = this.limits.get(identifier);
    
    if (!entry) {
      return {
        count: 0,
        limit: this.config.requestsPerWindow,
        remaining: this.config.requestsPerWindow,
        resetTime: Date.now() + this.config.windowMs,
        queueLength: 0
      };
    }

    const remaining = Math.max(0, this.config.requestsPerWindow - entry.count);
    const resetTime = entry.window + this.config.windowMs;

    return {
      count: entry.count,
      limit: this.config.requestsPerWindow,
      remaining,
      resetTime,
      queueLength: entry.queue.length
    };
  }

  /**
   * Get DDoS metrics
   */
  getDDoSMetrics(): DDoSMetrics {
    return {
      requestsPerSecond: this.ddosMetrics.requestsPerSecond,
      uniqueIPs: new Set(this.ddosMetrics.uniqueIPs),
      suspiciousPatterns: this.ddosMetrics.suspiciousPatterns,
      blockedRequests: this.ddosMetrics.blockedRequests
    };
  }

  /**
   * Reset rate limit for identifier
   */
  reset(identifier: string): void {
    this.limits.delete(identifier);
  }

  /**
   * Reset all rate limits
   */
  resetAll(): void {
    this.limits.clear();
    this.ddosMetrics.uniqueIPs.clear();
    this.ddosMetrics.requestsPerSecond = 0;
    this.ddosMetrics.suspiciousPatterns = 0;
    this.ddosMetrics.blockedRequests = 0;
  }

  /**
   * Update rate limit configuration
   */
  updateConfig(newConfig: Partial<RateLimitConfig>): void {
    Object.assign(this.config, newConfig);
    this.emit('configUpdated', this.config);
  }

  /**
   * Shutdown rate limiter
   */
  shutdown(): void {
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval);
    }
    
    // Process any remaining queued requests
    for (const [identifier] of this.limits) {
      this.processQueue(identifier);
    }
  }

  private getIdentifier(request: any): string {
    // Use IP address if available, otherwise use a general identifier
    return request.ip || request.clientId || request.headers?.['x-forwarded-for'] || 'default';
  }

  private isDDoSRequest(request: any, identifier: string): boolean {
    const now = Date.now();
    const entry = this.limits.get(identifier);
    
    // Check if requests are coming too fast
    if (entry && (now - entry.lastRequest) < 10) {
      return true;
    }

    // Check overall system load
    if (this.ddosMetrics.requestsPerSecond > this.config.ddosThreshold) {
      return true;
    }

    // Check for patterns typical of DDoS attacks
    const userAgent = request.headers?.['user-agent'];
    if (!userAgent || userAgent.length < 10) {
      this.ddosMetrics.suspiciousPatterns++;
      return this.ddosMetrics.suspiciousPatterns > 50;
    }

    return false;
  }

  private updateDDoSMetrics(request: any, identifier: string): void {
    this.ddosMetrics.requestsPerSecond++;
    
    const ip = this.extractIP(request);
    if (ip) {
      this.ddosMetrics.uniqueIPs.add(ip);
    }
  }

  private extractIP(request: any): string | null {
    return (
      request.ip ||
      request.connection?.remoteAddress ||
      request.headers?.['x-forwarded-for']?.split(',')[0] ||
      request.headers?.['x-real-ip'] ||
      null
    );
  }

  private monitorDDoS(): void {
    // Reset per-second counter
    this.ddosMetrics.requestsPerSecond = 0;
    
    // Clear old IPs (keep only last minute)
    if (this.ddosMetrics.uniqueIPs.size > 1000) {
      this.ddosMetrics.uniqueIPs.clear();
    }
  }

  private async processQueue(identifier: string): Promise<void> {
    const entry = this.limits.get(identifier);
    if (!entry || entry.queue.length === 0) {
      return;
    }

    const now = Date.now();
    
    // Process requests that can fit in the new window
    while (entry.queue.length > 0 && entry.count < this.config.requestsPerWindow) {
      const queuedRequest = entry.queue.shift()!;
      
      // Check if request hasn't timed out (5 minute timeout)
      if (now - queuedRequest.timestamp > 5 * 60 * 1000) {
        queuedRequest.reject(new Error('Request timeout'));
        continue;
      }

      entry.count++;
      queuedRequest.resolve(true);
    }

    // Reject remaining requests if queue is still full
    while (entry.queue.length > 0) {
      const queuedRequest = entry.queue.shift()!;
      queuedRequest.reject(new Error('Rate limit exceeded'));
    }
  }

  private cleanup(): void {
    const now = Date.now();
    const expiredThreshold = now - (this.config.windowMs * 2); // Keep for 2 windows
    
    for (const [identifier, entry] of this.limits.entries()) {
      if (entry.window < expiredThreshold && entry.queue.length === 0) {
        this.limits.delete(identifier);
      }
    }
  }
}

File: sdks/python/hypersim_sdk/security/rate_limiter.py
"""
Rate limiting with DDoS protection for Python SDK
"""

import asyncio
import time
from typing import Any, Dict, List
from .types import RateLimitConfig


class RateLimiter:
    """Rate limiting with DDoS protection"""
    
    def __init__(self, config: RateLimitConfig):
        self.config = config
        self.limits: Dict[str, Dict] = {}
        self.ddos_metrics = {
            'requests_per_second': 0,
            'unique_ips': set(),
            'suspicious_patterns': 0,
            'blocked_requests': 0
        }
        
        # Start cleanup and monitoring tasks
        asyncio.create_task(self._cleanup_task())
        if config.ddos_protection:
            asyncio.create_task(self._monitor_ddos())
    
    async def check_limit(self, request: Dict[str, Any]) -> bool:
        """Check if request is within rate limit"""
        identifier = self._get_identifier(request)
        now = int(time.time() * 1000)
        
        # Check for DDoS patterns
        if self.config.ddos_protection and self._is_ddos_request(request, identifier):
            self.ddos_metrics['blocked_requests'] += 1
            return False
        
        if identifier not in self.limits:
            self.limits[identifier] = {
                'count': 0,
                'window': now,
                'queue': [],
                'suspicious_activity': 0,
                'last_request': 0
            }
        
        entry = self.limits[identifier]
        
        # Reset window if expired
        if now - entry['window'] >= self.config.window_ms:
            entry['count'] = 0
            entry['window'] = now
            entry['suspicious_activity'] = 0
        
        # Check for suspicious patterns
        time_since_last = now - entry['last_request']
        if time_since_last < 100:  # Less than 100ms between requests
            entry['suspicious_activity'] += 1
            if entry['suspicious_activity'] > 10:
                return False
        
        entry['last_request'] = now
        
        # Check rate limit
        if entry['count'] >= self.config.requests_per_window:
            if (self.config.max_queue_size > 0 and 
                len(entry['queue']) < self.config.max_queue_size):
                # Queue the request
                future = asyncio.get_event_loop().create_future()
                entry['queue'].append({
                    'request': request,
                    'future': future,
                    'timestamp': now
                })
                return await future
            return False
        
        entry['count'] += 1
        self._update_ddos_metrics(request, identifier)
        return True
    
    def get_status(self, identifier: str) -> Dict:
        """Get current rate limit status for identifier"""
        entry = self.limits.get(identifier)
        
        if not entry:
            return {
                'count': 0,
                'limit': self.config.requests_per_window,
                'remaining': self.config.requests_per_window,
                'reset_time': int(time.time() * 1000) + self.config.window_ms,
                'queue_length': 0
            }
        
        remaining = max(0, self.config.requests_per_window - entry['count'])
        reset_time = entry['window'] + self.config.window_ms
        
        return {
            'count': entry['count'],
            'limit': self.config.requests_per_window,
            'remaining': remaining,
            'reset_time': reset_time,
            'queue_length': len(entry['queue'])
        }
    
    def get_ddos_metrics(self) -> Dict:
        """Get DDoS metrics"""
        return {
            'requests_per_second': self.ddos_metrics['requests_per_second'],
            'unique_ips': len(self.ddos_metrics['unique_ips']),
            'suspicious_patterns': self.ddos_metrics['suspicious_patterns'],
            'blocked_requests': self.ddos_metrics['blocked_requests']
        }
    
    def reset(self, identifier: str):
        """Reset rate limit for identifier"""
        self.limits.pop(identifier, None)
    
    def reset_all(self):
        """Reset all rate limits"""
        self.limits.clear()
        self.ddos_metrics = {
            'requests_per_second': 0,
            'unique_ips': set(),
            'suspicious_patterns': 0,
            'blocked_requests': 0
        }
    
    def _get_identifier(self, request: Dict[str, Any]) -> str:
        """Get request identifier for rate limiting"""
        return (request.get('ip') or 
                request.get('client_id') or 
                request.get('headers', {}).get('x-forwarded-for') or 
                'default')
    
    def _is_ddos_request(self, request: Dict[str, Any], identifier: str) -> bool:
        """Check if request shows DDoS patterns"""
        now = int(time.time() * 1000)
        entry = self.limits.get(identifier)
        
        # Check if requests are coming too fast
        if entry and (now - entry['last_request']) < 10:
            return True
        
        # Check overall system load
        if self.ddos_metrics['requests_per_second'] > self.config.ddos_threshold:
            return True
        
        # Check for suspicious user agents
        user_agent = request.get('headers', {}).get('user-agent', '')
        if not user_agent or len(user_agent) < 10:
            self.ddos_metrics['suspicious_patterns'] += 1
            return self.ddos_metrics['suspicious_patterns'] > 50
        
        return False
    
    def _update_ddos_metrics(self, request: Dict[str, Any], identifier: str):
        """Update DDoS monitoring metrics"""
        self.ddos_metrics['requests_per_second'] += 1
        
        ip = self._extract_ip(request)
        if ip:
            self.ddos_metrics['unique_ips'].add(ip)
    
    def _extract_ip(self, request: Dict[str, Any]) -> str:
        """Extract IP address from request"""
        return (request.get('ip') or
                request.get('headers', {}).get('x-forwarded-for', '').split(',')[0] or
                request.get('headers', {}).get('x-real-ip') or
                None)
    
    async def _cleanup_task(self):
        """Periodic cleanup of old entries"""
        while True:
            await asyncio.sleep(60)  # Run every minute
            now = int(time.time() * 1000)
            expired_threshold = now - (self.config.window_ms * 2)
            
            expired_keys = [
                key for key, entry in self.limits.items()
                if entry['window'] < expired_threshold and not entry['queue']
            ]
            
            for key in expired_keys:
                del self.limits[key]
    
    async def _monitor_ddos(self):
        """Monitor for DDoS attacks"""
        while True:
            await asyncio.sleep(1)  # Check every second
            
            # Reset per-second counter
            self.ddos_metrics['requests_per_second'] = 0
            
            # Clear old IPs (keep only last minute)
            if len(self.ddos_metrics['unique_ips']) > 1000:
                self.ddos_metrics['unique_ips'].clear()
